pkg/interceptor/rttfromxr.go

```diff
@@ -0,0 +1,192 @@
+	r.lock.Lock()
+	if !r.started {
+		r.started = true
+		go r.start()
+	}
+
+	r.remoteTrackSsrcs = append(r.remoteTrackSsrcs, info.SSRC)
+	r.lock.Unlock()
+
+	return reader
+}
+
+func (r *RTTFromXR) UnbindRemoteStream(info *interceptor.StreamInfo) {
+	r.lock.Lock()
+	for i, ssrc := range r.remoteTrackSsrcs {
+		if ssrc == info.SSRC {
+			r.remoteTrackSsrcs[i] = r.remoteTrackSsrcs[len(r.remoteTrackSsrcs)-1]
+			r.remoteTrackSsrcs = r.remoteTrackSsrcs[:len(r.remoteTrackSsrcs)-1]
+			break
+		}
+	}
+	r.lock.Unlock()
+}
+
+func (r *RTTFromXR) BindLocalStream(info *interceptor.StreamInfo, writer interceptor.RTPWriter) interceptor.RTPWriter {
+	r.lock.Lock()
+	r.localTrackSsrcs = append(r.localTrackSsrcs, info.SSRC)
+	r.lock.Unlock()
+	return writer
+}
+
+func (r *RTTFromXR) UnbindLocalStream(info *interceptor.StreamInfo) {
+	r.lock.Lock()
+	for i, ssrc := range r.localTrackSsrcs {
+		if ssrc == info.SSRC {
+			r.localTrackSsrcs[i] = r.localTrackSsrcs[len(r.localTrackSsrcs)-1]
+			r.localTrackSsrcs = r.localTrackSsrcs[:len(r.localTrackSsrcs)-1]
+			break
+		}
+	}
+	r.lock.Unlock()
+
+}
+
+func (r *RTTFromXR) BindRTCPReader(reader interceptor.RTCPReader) interceptor.RTCPReader {
+	return interceptor.RTCPReaderFunc(func(in []byte, a interceptor.Attributes) (int, interceptor.Attributes, error) {
+		n, a, err := reader.Read(in, a)
+		if err != nil {
+			return n, a, err
+		}
+
+		if a == nil {
+			a = make(interceptor.Attributes)
+		}
+		pkts, err := a.GetRTCPPackets(in[:n])
+		if err != nil {
+			return n, a, err
+		}
+
+		for _, pkt := range pkts {
+			if xr, ok := pkt.(*rtcp.ExtendedReport); ok {
+				r.handleXR(xr)
+			}
+		}
+		return n, a, nil
+	})
+}
+
+func (r *RTTFromXR) BindRTCPWriter(writer interceptor.RTCPWriter) interceptor.RTCPWriter {
+	r.lock.Lock()
+	r.writer = writer
+	r.lock.Unlock()
+	return writer
+}
+
+func (r *RTTFromXR) Close() error {
+	close(r.close)
+	return nil
+}
+
+func (r *RTTFromXR) handleXR(xr *rtcp.ExtendedReport) {
+	for _, report := range xr.Reports {
+		switch rr := report.(type) {
+		case *rtcp.ReceiverReferenceTimeReportBlock:
+			var dlrrs []rtcp.DLRRReport
+			r.lock.Lock()
+			for _, ssrc := range r.localTrackSsrcs {
+				dlrrs = append(dlrrs, rtcp.DLRRReport{
+					SSRC:   ssrc,
+					LastRR: uint32(rr.NTPTimestamp >> 16),
+				})
+			}
+			writer := r.writer
+			r.lock.Unlock()
+			if len(dlrrs) > 0 && writer != nil {
+				writer.Write([]rtcp.Packet{&rtcp.ExtendedReport{
+					SenderSSRC: xr.SenderSSRC,
+					Reports: []rtcp.ReportBlock{
+						&rtcp.DLRRReportBlock{
+							Reports: dlrrs,
+						},
+					},
+				}}, nil)
+			}
+
+		case *rtcp.DLRRReportBlock:
+			for _, dlrrReport := range rr.Reports {
+				nowNTP := util.ToNtpTime(time.Now())
+				nowNTP32 := uint32(nowNTP >> 16)
+				ntpDiff := nowNTP32 - dlrrReport.LastRR - dlrrReport.DLRR
+				rtt := uint32(math.Ceil(float64(ntpDiff) * 1000.0 / 65536.0))
```

Comment:
Should the send above set DLRR? Did not see it set on the send side? Did I miss it? Maybe it is set in some other interceptor?

Reply:
Maybe it is close to 0 because it is read and written back immediately? If so, a comment would be useful there.

Sorry getting confused reading the interceptor because local/remote, reader/writer, send/recv all in one place is messing with my head 🙂

Reply:
No DLRR since it responds immediately, commented